# Credit Risk Evaluator

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier

## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [2]:
# Import the data
init_df = pd.read_csv("Resources/lending_data.csv")

In [3]:
#View initial data
init_df.head(5)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [4]:
init_df.shape

(77536, 8)

In [5]:
#QC the dataframe
#Check for nulls
init_df.isnull().values.any()

False

In [6]:
#QC the dataframe cont...
#Check data type of columns
init_df.dtypes

loan_size           float64
interest_rate       float64
borrower_income       int64
debt_to_income      float64
num_of_accounts       int64
derogatory_marks      int64
total_debt            int64
loan_status           int64
dtype: object

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*I believe that logistic regression will perform better than random forest. The data set, at 77536 rows, is not particularly large. My understanding is the value of random forest, outside of processing time, is honing its preditions with large data sets. Also there are several potential variables, pertaining to loans and personal finance, that are not captured in data set, which further limits the size. \
*Also, all the data is numeric. Each column in the dataframe is either float or int. Broadly speaking logistic regression, vs random forest, works well with numbers. All we have is numbers in this data set.*

## Split the Data into Training and Testing Sets

In [7]:
# Split the data into X_train, X_test, y_train, y_test
x_loan = init_df.drop(columns=['loan_status'])
x_loan.head(5)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [8]:
y_loan = init_df['loan_status']
y_loan.head(5)

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [9]:
X_train, X_test, y_train, y_test = train_test_split(x_loan, y_loan, random_state=1)

## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

In [10]:
# Train a Logistic Regression model.
loan_LR = LogisticRegression(max_iter=15000)
loan_LR.fit(X_train, y_train)

LogisticRegression(max_iter=15000)

In [11]:
#Print of Logistic Regression model score.
print(f"LR_Training Score: {loan_LR.score(X_train, y_train)}")
print(f"LR_Testing Score: {loan_LR.score(X_test, y_test)}")

LR_Training Score: 0.9921240885954051
LR_Testing Score: 0.9918489475856377


In [12]:
# Train a Random Forest Classifier model and print the model score
scaler_loan = StandardScaler().fit(X_train)
X_train_scaled = scaler_loan.transform(X_train)
X_test_scaled = scaler_loan.transform(X_test)
loan_rfc = RandomForestClassifier(random_state=1, n_estimators=750).fit(X_train_scaled, y_train)

In [13]:
#Print of Logistic Regression model score.
print(f"RFC_Training Score: {loan_rfc.score(X_train_scaled, y_train)}")
print(f"RFC_Testing Score: {loan_rfc.score(X_test_scaled, y_test)}")

RFC_Training Score: 0.9975409272252029
RFC_Testing Score: 0.9917457697069748


*The scores are extremely close, with the biggest difference being between the training score of the two models. This does not agree with my prediction, where I indicated logistic regression would have a clear advantage. \
My thought is that the num_of_accounts and derogatory_marks are low, non decimal values that may act more like classifiers than numbers, but I would have to explore this much further. \
Either way, both models have acceptable accuracy scores.*